### PR TITLE
Fix CPython 3.11 compatibility

### LINKED ISSUE #43: Genshi doesn't work on Python 3.11: code expected at least 18 arguments, got 16
Example of error when running tests:

```
======================================================================
FAIL: OldTextTemplate (genshi.template.text)
Doctest: genshi.template.text.OldTextTemplate
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vstinner/python/main/Lib/doctest.py", line 2216, in runTest
    raise self.failureException(self.format_failure(new.getvalue()))
AssertionError: Failed doctest test for genshi.template.text.OldTextTemplate
  File "/home/vstinner/dev/genshi/genshi/template/text.py", line 239, in OldTextTemplate

----------------------------------------------------------------------
File "/home/vstinner/dev/genshi/genshi/template/text.py", line 244, in genshi.template.text.OldTextTemplate
Failed example:
    tmpl = OldTextTemplate('''Dear $name,

    We have the following items for you:
    #for item in items
     * $item
    #end

    All the best,
    Foobar''')
Exception raised:
    Traceback (most recent call last):
      File "/home/vstinner/python/main/Lib/doctest.py", line 1348, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest genshi.template.text.OldTextTemplate[0]>", line 1, in <module>
        tmpl = OldTextTemplate('''Dear $name,
      File "/home/vstinner/dev/genshi/genshi/template/base.py", line 421, in __init__
        self._stream = self._parse(source, encoding)
      File "/home/vstinner/dev/genshi/genshi/template/text.py", line 293, in _parse
        for kind, data, pos in interpolate(text, self.filepath, lineno,
      File "/home/vstinner/dev/genshi/genshi/template/interpolation.py", line 77, in interpolate
        expr = Expression(chunk.strip(), pos[0], pos[1],
      File "/home/vstinner/dev/genshi/genshi/template/eval.py", line 71, in __init__
        self.code = _compile(node, self.source, mode=self.mode,
      File "/home/vstinner/dev/genshi/genshi/template/eval.py", line 453, in _compile
        return build_code_chunk(code, filename, name, lineno)
      File "/home/vstinner/dev/genshi/genshi/compat.py", line 110, in build_code_chunk
        return CodeType(*params)
    TypeError: code() argument 15 must be bytes, not tuple
----------------------------------------------------------------------
File "/home/vstinner/dev/genshi/genshi/template/text.py", line 253, in genshi.template.text.OldTextTemplate
Failed example:
    print(tmpl.generate(name='Joe', items=[1, 2, 3]).render(encoding=None))
Exception raised:
    Traceback (most recent call last):
      File "/home/vstinner/python/main/Lib/doctest.py", line 1348, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest genshi.template.text.OldTextTemplate[1]>", line 1, in <module>
        print(tmpl.generate(name='Joe', items=[1, 2, 3]).render(encoding=None))
    NameError: name 'tmpl' is not defined
```

### LINKED ISSUE #52: CodeType: swapped parameter order for co_kwonlyargcount and co_nlocals
initially reported by @ztane in https://github.com/edgewall/genshi/issues/43#issuecomment-955814422

> Since I was doing this for my own templating engine [Tonnikala](https://github.com/tetframework/tonnikala) and was just casually browsing around, I noticed that Genshi seems to have accidentally swapped `co_kwonlyargcount` and `co_nlocals` c.f. CPython help() for code objects...

I think this does not affect Genshi on Python 3.8+ as we are using `.replace()` there but it still is a bug in Genshi.

Current Genshi code for Python 3: https://github.com/edgewall/genshi/blob/a23f3054b96b487215b04812c680075c5117470a/genshi/compat.py#L109-L113

Actual constructor arguments in CPython 3.5:
https://github.com/python/cpython/blob/7544508f0245173bff5866aa1598c8f6cce1fc5f/Lib/types.py#L60-L65

Constructor for CPython 3.11 referenced in https://github.com/edgewall/genshi/issues/43#issuecomment-845491080

Genshi code Python 2: https://github.com/edgewall/genshi/blob/a23f3054b96b487215b04812c680075c5117470a/genshi/compat.py#L91-L94

Maybe I'm just confused right now but it seems the Python 2 constructor does not use the correct ordering as well: https://github.com/python/cpython/blob/3bbd2fad4d4a282c7a5a3169a4f497b97aeff319/Objects/codeobject.c#L50-L56

For Python 3 the constructor using positional parameters only is marked as "deprecated" so maybe we should use keyword parameters only.

### PR DESCRIPTION
**Pull request by brandtbucher**
Fixes #43. Fixes #52.

### CODE REVIEW

File: genshi/compat.py
Comment (brandtbucher) [inline genshi/compat.py:116]: This attribute is never present here, so I removed the branch and cleaned things up a bit.

File: genshi/compat.py
Comment (brandtbucher) [inline genshi/compat.py:131]: Thoughts?

**Comment by brandtbucher** [comment]
Hm, looks like CI isn't configured?

I ran the tests locally using 3.7, 3.8, 3.9, 3.10, and 3.11. All pass.

**Comment by brandtbucher** [comment]
CC @FelixSchwarz.

This issue is breaking [`pyperformance`](https://github.com/python/pyperformance) on CPython's `main` branch; happy to help out any way I can to get it back up and running!

**Comment by FelixSchwarz** [comment]
Thank you so much for this PR, looks good to me. I'm sorry that it took so long until you got a response but I'll try to help you getting this landed (unless Simon beats me to it).

> Hm, looks like CI isn't configured?

CI might be restricted to actual commits in the main repo as it contains deployment keys or so. @hodgestar knows more about this.

File: genshi/compat.py
Comment (hodgestar) [reply genshi/compat.py:131]: `build_code_chunk` is only called in `genshi/template/eval.py` with the following note:
```
    try:
        # We'd like to just set co_firstlineno, but it's readonly. So we need
        # to clone the code object while adjusting the line number
        return build_code_chunk(code, filename, name, lineno)
    except RuntimeError:
        return code
```
Based on the above note, I added the `.replace(...)` version of `build_code_chunk` fairly recently (well, post 3.8) and I think one day maybe the older version can just be removed and we can just call `code.replace` directly.

**Review by hodgestar — APPROVED** [review]

**Comment by hodgestar** [comment]
@brandtbucher This looks good -- thank you!

The CI tests are not running because I forgot to put a `, pull_request` in `on: [pull]` in the GitHub actions file. It's been added in #57. We can either wait a bit for #57 to land and then merge master into this PR or just add `, pull_request` here too and deal with the merge conflict later.

I'd like to run tests on CI before merging to avoid anything odd happening.

**Comment by hodgestar** [comment]
@brandtbucher -- #57 has been merged, so if you can merge master into your branch, then I can run the CI tests on this PR.

**Comment by brandtbucher** [comment]
There you go! Just need maintainer approval to run CI.

**Comment by hodgestar** [comment]
@brandtbucher Thank you! Tests passed in CI. Merging.

**Comment by brandtbucher** [comment]
Any idea when we can expect a release with this support included, @hodgestar?

**Comment by FelixSchwarz** [comment]
> Any idea when we can expect a release with this support included, @hodgestar?

Yeah, having a Python 3.11 compatible release would be nice as this would also simplify Fedora 36 packaging. I think F36 will start a Python 3.11 mass rebuild in February.

**Pull request merged**
